### ISSUE TITLE
Pyright cannot do typing logic with union of types correctly

### ISSUE THREAD

**Issue by acampove** [microsoft/pyright#10748]
**Describe the bug**

Hi,

I am doing:

```python
var : list[str]|None = None

if var is None:
    return False

if x not in var:
    return True
```

And I keep getting:

```
Value `var` doesn't support membership test (unsupported-membership-test)
```

the first if is already ensuring that the variable is not None so it can only be a list and the error should not happen.

**Comment by erictraut** [comment]
Could you provide a complete code sample that demonstrates the problem you're seeing? The code sample above is incomplete in that it has syntax errors and references variables that are not defined. If I modify it to eliminate these problems, it type checks without problems, so there's likely some other issue in your code that's at fault here.

```python
def func(var: list[str] | None, x: str):
    if var is None:
        return False

    if x not in var:
        return True
```

I'm also switching this from a bug report to a discussion topic since this is unlikely to be a bug in pyright but rather a question about how type checking works.
